File: src/leetCode/easy_14_longestCommonPrefix/longestCommonPrefix.ts
// https://leetcode-cn.com/problems/longest-common-prefix/

/**
 * @param {string[]} strs
 * @returns {string}
 */
function longestCommonPrefixA(strs: string[]): string {
  if (strs.length === 0) {
    return "";
  }

  let prefix: string = strs[0];

  for (let i = 1; i < strs.length; i++) {
    prefix = lcp(prefix, strs[i]);

    if (prefix.length === 0) {
      return "";
    }
  }

  return prefix;
}

/**
 * @param {string} str1
 * @param {string} str2
 * @returns {string}
 */
function lcp(str1: string, str2: string): string {
  const minLength = Math.min(str1.length, str2.length);

  let index: number = 0;

  while (index < minLength && str1[index] === str2[index]) {
    index += 1;
  }

  return str1.substring(0, index);
}

/**
 * @param {string[]} strs
 * @returns {string}
 */
function longestCommonPrefixA1(strs: string[]): string {
  if (strs.length === 0) {
    return "";
  }

  let minIndex: number = strs[0].length - 1;

  for (let i = 0; i < strs.length; i++) {
    let j = 0;

    while (j <= minIndex) {
      if (j > strs[i].length - 1 || strs[0][j] !== strs[i][j]) {
        if (j === 0) {
          return "";
        }
        break;
      }

      j += 1;
    }
    minIndex = j - 1;
  }

  return strs[0].substring(0, minIndex + 1);
}

/**
 * @param {string[]} strs
 * @returns {string}
 */
function longestCommonPrefixA2(strs: string[]): string {
  if (strs.length === 0) return '';

  let minLength: number = strs[0].length;

  for (let i = 1; i < strs.length; i++) {
    if (strs[i].length < minLength) {
      minLength = strs[i].length;
    }
  }

  let prefix: string = strs[0].substr(0, minLength);

  for (let i = 1; i < strs.length; i++) {
    let j = 0;

    for (; j < minLength; j++) {
      if (prefix[j] !== strs[i][j]) {
        break;
      }
    }

    prefix = prefix.substring(0, j);

    if (!prefix) {
      return '';
    }
  }

  return prefix;
};

/**
 * @param {string[]} strs
 * @returns {string}
 */
function longestCommonPrefixB(strs: string[]): string {
  if (strs.length === 0) return '';

  const str0Length: number = strs[0].length;

  for (let i = 0; i < str0Length; i++) {
    for (let j = 0; j < strs.length; j++) {
      if (i > strs[j].length - 1 || strs[j][i] !== strs[0][i]) {
        return strs[0].substring(0, i);
      }
    }
  }

  return strs[0];
}

/**
 * @param {string[]} strs
 * @returns {string}
 */
function longestCommonPrefixC(strs: string[]): string {
  if (strs.length === 0) return '';

  function lcp(start: number, end: number): string {
    if (start === end) {
      return strs[start];
    }

    const mid = (start + end) / 2;

    const lcpLeft: string = lcp(start, mid);
    const lcpRight: string = lcp(mid + 1, end);

    const minLength = Math.min(lcpLeft.length, lcpRight.length);

    for (let i = 0; i < minLength; i++) {
      if (lcpLeft[i] !== lcpRight[i]) {
        return lcpLeft.substring(0, i);
      }
    }

    return lcpLeft.substring(0, minLength);
  }

  return lcp(0, strs.length - 1);
}

/**
 * @param {string[]} strs
 * @returns {string}
 */
function longestCommonPrefixD(strs: string[]): string {
  if (strs.length === 0) return '';

  function isCommonPrefix(len: number): boolean {
    for (let i = 0; i < strs.length; i++) {
      const subStr0 = strs[0].substring(0, len);
      const subStrI = strs[i].substring(0, len);

      if (subStr0 !== subStrI) {
        return false;
      }
    }

    return true;
  }

  let minLength = strs[0].length;

  strs.forEach(str => {
    if (str.length < minLength) {
      minLength = str.length;
    }
  });

  let left = 0, right = minLength;

  while (left < right) {
    const mid = left + (right - left + 1) / 2;

    if (isCommonPrefix(mid)) {
      left = mid;
    } else {
      right = mid - 1;
    }
  }

  return strs[0].substring(0, left);
}

File: src/leetCode/easy_14_longestCommonPrefix/longestCommonPrefix.go
package longestcommonprefix

import "strings"

// https://leetcode-cn.com/problems/longest-common-prefix/

// 解法1：横向扫描法(standard)
// 时间复杂度：O(mn)
// 空间复杂度：O(1)
func longestCommonPrefixA(strs []string) string {
	if len(strs) == 0 {
		return ""
	}

	prefix := strs[0]

	for i := 1; i < len(strs); i++ {
		prefix = lcpA(prefix, strs[i])

		if len(prefix) == 0 {
			return ""
		}
	}

	return prefix
}

func lcpA(str1, str2 string) string {
	minLength := min(len(str1), len(str2))
	i := 0

	for i < minLength && str1[i] == str2[i] {
		i++
	}

	return str1[:i]
}

func min(x, y int) int {
	if x < y {
		return x
	}

	return y
}

// 解法1.1：横向扫描法(mutation)
func longestCommonPrefixA1(strs []string) string {
	if len(strs) == 0 {
		return ""
	}

	minIndex := len(strs[0]) - 1

	for i := 0; i < len(strs); i++ {
		j := 0

		for j <= minIndex {
			// failed case
			if j > len(strs[i])-1 || strs[0][j] != strs[i][j] {
				if j == 0 {
					return ""
				}
				break
			}
			j++
		}

		minIndex = j - 1
	}

	return strs[0][:minIndex+1]
}

// 解法1.2：横向扫描法(mutation)
func longestCommonPrefixA2(strs []string) string {
	if len(strs) == 0 {
		return ""
	}

	prefix := strs[0]

	for i := 1; i < len(strs); i++ {
		for strings.Index(strs[i], prefix) != 0 {
			if len(prefix) == 0 {
				return ""
			}

			prefix = prefix[:len(prefix)-1]
		}
	}

	return prefix
}

// ============================================================================

// 解法2：纵向扫描法
// 时间复杂度：O(mn)
// 空间复杂度：O(1)
func longestCommonPrefixB(strs []string) string {
	if len(strs) == 0 {
		return ""
	}
	for i := 0; i < len(strs[0]); i++ {
		for j := 1; j < len(strs); j++ {
			last := len(strs[j]) - 1

			if i > last || strs[j][i] != strs[0][i] {
				return strs[0][:i]
			}
		}
	}
	return strs[0]
}

// ============================================================================

// 解法3：分治法
// 时间复杂度：O(mn)
// 空间复杂度：O(mlogn)
func longestCommonPrefixC(strs []string) string {
	if len(strs) == 0 {
		return ""
	}

	var lcp func(int, int) string

	lcp = func(start, end int) string {
		if start == end {
			return strs[start]
		}

		mid := (start + end) / 2

		lcpLeft, lcpRight := lcp(start, mid), lcp(mid+1, end)

		minLength := min(len(lcpLeft), len(lcpRight))

		for i := 0; i < minLength; i++ {
			if lcpLeft[i] != lcpRight[i] {
				return lcpLeft[:i]
			}
		}
		return lcpLeft[:minLength]
	}

	return lcp(0, len(strs)-1)
}

// ============================================================================

// 解法4：二分法
// 时间复杂度：O(mnlogm)
// 空间复杂度：O(1)
func longestCommonPrefixD(strs []string) string {
	if len(strs) == 0 {
		return ""
	}

	isCommonPrefix := func(length int) bool {
		for i := 1; i < len(strs); i++ {
			if strs[i][:length] != strs[0][:length] {
				return false
			}
		}

		return true
	}

	minLength := len(strs[0])

	for _, str := range strs {
		if len(str) < minLength {
			minLength = len(str)
		}
	}

	left, right := 0, minLength

	for left < right {
		mid := left + (right-left+1)/2 // 右中位数

		if isCommonPrefix(mid) {
			left = mid
		} else {
			right = mid - 1
		}
	}

	return strs[0][:left]
}
